### PR TITLE
Fix non reflective property of constrained hilbert spaces

### LINKED ISSUE #1374: Issues implementing metropolis on a Hamiltonian
I'm trying to train a model using NN. My Hamiltonian is a Hubbard model with varying on-site interaction (not constant V). So I wrote a customized Hamiltonian as per my need and input it for training a NN given in [this](https://netket.readthedocs.io/en/latest/tutorials/gs-matrix-models.html?highlight=eigenvalue#neural-network-quantum-state) tutorial. However, I am getting errors in training this Hamiltonian, saying it is an incompatible class for training. Can anyone please suggest a workaround for this and some info about what is going on? Attaching the relevant codes and errors:

* The Hamiltonian
```
Lambda = 4  # cutoff of each bosonic Fock space
N = 9  # number of bosons in our system 
L = 11 # no of sites in our bosonic system

hi = nk.hilbert.Fock(n_max = Lambda, N = L, n_particles = N)  # n_max -> Maximum occupation for a site

#the system parameters

V = 1.0
J = 1.0
U = 1.0

from netket.operator.boson import create, destroy, number

# we loop over all "sites" -> loop over N bosons
H_hop = sum([-J * (create(hi, i) * destroy(hi, i+1) + create(hi, i+1) * destroy(hi, i)) for i in range(N)])

H_onsite = sum([V * (i - L//2)**2 * (create(hi, i) * destroy(hi, i)) for i in range(L)])

H_interact = sum([U/2 * number(hi, i)*(number(hi, i) - 1) for i in range(N)])

H = H_hop + H_onsite + H_interact

eig_vals = nk.exact.lanczos_ed(
  H, k=13, compute_eigenvectors=False
)
```
* The error in training:

```
# trying out a different MLP variant
class MLP(flax.linen.Module):
  n_hidden_units: int

  @flax.linen.compact
  def __call__(self, x):
    
    W_0 = flax.linen.Dense(features=self.n_hidden_units, 
                    dtype=np.complex128,  
                    kernel_init=jax.nn.initializers.normal(stddev=0.1), 
                    bias_init=jax.nn.initializers.normal(stddev=0.1))(x)
    
    #W_0 = nn.activation.tanh(W_0)

    W_1 = flax.linen.Dense(features=2, 
                      dtype=np.complex128, 
                      kernel_init=jax.nn.initializers.normal(stddev=0.1), 
                      bias_init=jax.nn.initializers.normal(stddev=0.1))(W_0)
        
    y = flax.linen.tanh(W_1)
    return jnp.sum(y,axis = -1)

model = MLP(n_hidden_units = 20)

# Create the an exact sampler over the Hilber space , if it is not too large
sampler = nk.sampler.MetropolisLocal(H, n_chains = 6)

# create the VMC state
vstate = nk.vqs.MCState(sampler, model, n_samples=2**10)

optimizer = nk.optimizer.Sgd(learning_rate=0.05)
preconditioner = nk.optimizer.SR(diag_shift=0.1)

# Notice the use, again of Stochastic Reconfiguration, which considerably improves the optimisation
gs = nk.driver.VMC(
    H, optimizer, variational_state=vstate, preconditioner=preconditioner
)

# logging and running : sometimes need to rung longer than 300 to exit a plateaux
log = nk.logging.RuntimeLog()
gs.run(n_iter=300, out=log)

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
[<ipython-input-55-f6ba8b9948e2>](https://localhost:8080/#) in <module>
     24 
     25 # Create the an exact sampler over the Hilber space , if it is not too large
---> 26 sampler = nk.sampler.MetropolisLocal(H, n_chains = 6)
     27 
     28 # create the VMC state

4 frames
[/usr/local/lib/python3.7/dist-packages/netket/sampler/base.py](https://localhost:8080/#) in __post_init__(self)
    119             raise ValueError(
    120                 "hilbert must be a subtype of netket.hilbert.AbstractHilbert, "
--> 121                 + "instead, type {} is not.".format(type(self.hilbert))
    122             )
    123 

ValueError: hilbert must be a subtype of netket.hilbert.AbstractHilbert, instead, type <class 'netket.operator._local_operator.LocalOperator'> is not.
```

### PR DESCRIPTION
**Pull request by PhilipVinc**
Also clarifies a few error messages in simpers .

reported in https://github.com/netket/netket/issues/1374#issuecomment-1295864438
fixes #1374

### CODE REVIEW

File: test/hilbert/test_hilbert.py [hunk 1]
```diff
@@ -519,3 +519,15 @@ def test_tensor_combination():
...
     assert len(hit._hilbert_spaces) == 5
     repr(hit)
+
+
+def test_constrained_equivalence():
+    hi1 = nk.hilbert.Spin(0.5, 4, total_sz=0)
+    hi2 = nk.hilbert.Spin(0.5, 4, total_sz=0)
+    assert hi1 == hi2
+    assert hash(hi1) == hash(hi2)
+
```
Comment (gcarleo) [inline test/hilbert/test_hilbert.py:529]: maybe to be super safe we could add a check 

```python
    hi1 = nk.hilbert.Spin(0.5, 4, total_sz=0)
    hi2 = nk.hilbert.Spin(0.5, 4, total_sz=2)

    assert not has(h1) == hash(hi2) 
    
```

**Review by gcarleo — APPROVED** [review]
ginger beer

**Comment by gcarleo** [comment]
tavernello

**Comment by PhilipVinc** [comment]
beautiful https://netket--1375.org.readthedocs.build/en/1375/api/_generated/optim/netket.optimizer.SR.html#netket.optimizer.SR

**Pull request merged**